addons/l10n_pt_pos/models/pos_order.py

```diff
@@ -0,0 +1,220 @@
+
+    country_code = fields.Char(related='company_id.account_fiscal_country_id.code', depends=['company_id.account_fiscal_country_id'])
+    l10n_pt_pos_document_number = fields.Char(string='Portuguese Document Number', compute='_compute_l10n_pt_pos_document_number', store=True)
+    l10n_pt_pos_inalterable_hash = fields.Char(string="Inalterability Hash", readonly=True, copy=False)
+    l10n_pt_pos_qr_code_str = fields.Char(string='Portuguese QR Code', compute='_compute_l10n_pt_pos_qr_code_str', store=True)
+    l10n_pt_pos_inalterable_hash_short = fields.Char(string='Short version of the Portuguese hash', compute='_compute_l10n_pt_pos_inalterable_hash_short')
+    l10n_pt_pos_atcud = fields.Char(string='Portuguese ATCUD', compute='_compute_l10n_pt_pos_atcud', store=True)
+
+    @api.depends('name', 'company_id.account_fiscal_country_id.code')
+    def _compute_l10n_pt_pos_document_number(self):
+        for order in self.filtered(lambda o: (
+            o.company_id.account_fiscal_country_id.code == 'PT'
+            and o.config_id.l10n_pt_pos_official_series_id
+            and o.name
+            and not o.l10n_pt_pos_document_number
+        )):
+            sequence_prefix, sequence_number = order._l10n_pt_pos_get_sequence_info()
+            order.l10n_pt_pos_document_number = f"pos_order {sequence_prefix}/{sequence_number}"
+
+    @api.depends('l10n_pt_pos_inalterable_hash')
+    def _compute_l10n_pt_pos_inalterable_hash_short(self):
+        for order in self:
+            order.l10n_pt_pos_inalterable_hash_short = L10nPtHashingUtils._l10n_pt_get_short_hash(order.l10n_pt_pos_inalterable_hash)
+
+    @api.depends('name', 'config_id.l10n_pt_pos_official_series_id.code', 'l10n_pt_pos_inalterable_hash')
+    def _compute_l10n_pt_pos_atcud(self):
+        for order in self:
+            if (
+                order.company_id.account_fiscal_country_id.code == 'PT'
+                and order.config_id.l10n_pt_pos_official_series_id
+                and order.l10n_pt_pos_inalterable_hash
+                and not order.l10n_pt_pos_atcud
+            ):
+                order.l10n_pt_pos_atcud = f"{order.config_id.l10n_pt_pos_official_series_id.code}-{order._l10n_pt_pos_get_sequence_info()[1]}"
+            else:
+                order.l10n_pt_pos_atcud = False
+
+    @api.depends('l10n_pt_pos_atcud')
+    def _compute_l10n_pt_pos_qr_code_str(self):
+        """ Generate the informational QR code for Portugal invoicing.
+        E.g.: A:509445535*B:123456823*C:BE*D:FT*E:N*F:20220103*G:FT 01P2022/1*H:0*I1:PT*I7:325.20*I8:74.80*N:74.80*O:400.00*P:0.00*Q:P0FE*R:2230
+        """
+
+        def format_amount(pos_order, amount):
+            """
+            Convert amount to EUR based on the rate of a given account_move's date
+            Format amount to 2 decimals as per SAF-T (PT) requirements
+            """
+            amount_eur = pos_order.currency_id._convert(amount, self.env.ref('base.EUR'), pos_order.company_id, pos_order.date_order)
```

Comment:
Don't `ref()` in a loop, it executes a `exists()` every time (one useless query easily avoidable)